File: ProblemSolve/221.cpp
/******************
 * Maximal Square *
 ******************/

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int len = matrix.size();
        int len2 = matrix[0].size();        
        int ans = 0;
        vector<vector<int>> dp(len + 1, vector<int>(len2 + 1, 0));
        
        for(int i=1; i<=len; i++){
            for(int j=1; j<=len2; j++){
                if(matrix[i-1][j-1] == '1'){                    
                    //dp[i][j]는 (i,j)를 우측 하단 꼭지점으로하는 정사각형의 변의 길이를 나타낸다
                    //왼쪽칸, 왼쪽위칸, 위칸을 비교하면서 가능한 사각형의 크기를 설정
                    dp[i][j] = min(dp[i-1][j], min(dp[i-1][j-1], dp[i][j-1])) + 1;
                }                
                
                //가능한 정사각형의 가장 큰 변의 길이를 저장
                ans = max(ans, dp[i][j]);                
            }
        }
        
        return ans * ans;
    }
};

File: ProblemSolve/221.cs
/******************
 * Maximal Square *
 ******************/

public class Solution {
    public int MaximalSquare(char[][] matrix) {
        int lenRow = matrix.Length;
        int lenCol = matrix[0].Length;
        int ans = 0;
    
        int[,] dp = new int[lenRow + 1,lenCol + 1];
        //인접한 3곳의 값을 확인해 가장 작은 값에 + 1을 하면 만들 수 있는 정사각형의 한변의 길이가 된다.
        //ex) matrix ={{1, 1}, {1, 1}} -> dp = {{0, 0, 0}, {0, 1, 1}, {0, 1, 2}}
        for(int i=1; i<=lenRow; ++i){
            for(int j=1; j<=lenCol; ++j){
                if(matrix[i-1][j-1] == '1'){
                    dp[i,j] = Math.Min(dp[i-1,j], Math.Min(dp[i,j-1], dp[i-1,j-1])) + 1;
                }
                ans = Math.Max(ans, dp[i,j]);
            }
        }
        
        return ans * ans;
    }
}
